### DISCUSSION TITLE
Support for torch.nn.ConstantPad2d / ONNX Pad in hls4ml

### DISCUSSION THREAD

**Discussion by The-Padi** [fastmachinelearning/hls4ml#1318] [Ideas]
Hi everyone,

I'd like to suggest support for torch.nn.ConstantPad2d in hls4ml.

This module is a simple and explicit way to apply zero-padding in PyTorch models. It's clean to use, exportable to ONNX without issues, and especially useful in modular or quantized models where padding is applied outside the convolution layer (rather than as padding= in Conv2d).

Currently, when exporting a model using ConstantPad2d, the ONNX graph contains a Pad node with a pads attribute and a constant value (usually 0). However, hls4ml doesn't support Pad yet, even though it already has internal support for ZeroPadding2D (as used by Keras models).

Would it be feasible to map ONNX Pad nodes with constant value 0 to ZeroPadding2D in hls4ml? I believe this would improve support for PyTorch models and allow more flexibility in network architectures.

Thanks for your time!

**Comment by vloncar** [comment]
This should be fairly simple, let me see if I can find a volunteer to implement it.

**Comment by vloncar** [comment]
Resolved in #1318

**Comment by The-Padi** [comment]
I just did a few test. This is not resolved: "ConstantPad" support was added for pytorch (009bcf5a28a6c0e6d569a63e67e31c6e5c0a364c) converter but not the "Pad" layer of ONNX/QONNX. @vloncar could this be re-opened ?

**Reply by vloncar** [reply]
yeah, we'll get to this too, i thought you needed one or the other not both.

**Reply by The-Padi** [reply]
Sorry for the misunderstanding. The ONNX/QONNX support was what I was really looking for.

**Reply by vloncar** [reply]
#1352
